src/main/java/com/ncedu/fooddelivery/api/v1/services/impls/OrderServiceImpl1.java

```diff
@@ -32,20 +81,493 @@ public Order getOrder(Long id) {
+        if(user.getRole() == Role.CLIENT) return orderRepo.getClientOrdersHistory(user.getId(), pageable).stream()
+                                                          .map(order -> convertToOrderInfoDTO(order))
+                                                          .collect(Collectors.toList());
+
+        return orderRepo.getCourierOrdersHistory(user.getId(), pageable).stream()
+                        .map(order -> convertToOrderInfoDTO(order)).collect(Collectors.toList());
+    }
+
+    @Override
+    public Double[] countOrderCost(CoordsDTO geo,
+                                   HashMap<Long, Integer> pairs, Long clientWarehouseId) {
+        WarehouseInfoDTO warehouse = warehouseService.getNearestWarehouse(geo.getLat(), geo.getLon());
+        if(warehouse == null) throw new NotFoundEx(String.format("{Lat: %s; lon: %s}", geo.getLat().toString(),
+                geo.getLon().toString()));
+        Long warehouseId = warehouse.getId();
+        if(!warehouseId.equals(clientWarehouseId)) throw new WarehouseCoordsBindingEx();
+
+        Map<Long, Integer> productsAvailableAmounts = new HashMap<>();
+
+        Double overallOrderCost = 0.0;
+        Double overallOrderDiscount = 0.0;
+        Double orderHighDemandCoeff;
+
+        for(Map.Entry<Long, Integer> pair: pairs.entrySet()){
+            Long productId = pair.getKey();
+            Integer requestedAmount = pair.getValue();
+
+            // 'findByProductIdAndWarehouseIdWithLock' sets lock on every product position in result set. So no one
+            // other thread
+            // will be able to interact with these positions before whole transaction will complete.
+            List<ProductPosition> productPositions = productPositionRepo.findByProductIdAndWarehouseIdWithLock(productId, warehouseId);
+
+            //filtering expired product positions
+            productPositions = productPositions.stream().filter(new Predicate<ProductPosition>() {
+                @Override
+                public boolean test(ProductPosition productPosition) {
+                    Short expirationDays = productPosition.getProduct().getExpirationDays();
+                    Date manufactureDate = productPosition.getManufactureDate();
+                    Calendar c = Calendar.getInstance();
+                    c.setTime(manufactureDate);
+                    c.add(Calendar.DATE , expirationDays);
+                    if(c.before(Calendar.getInstance())) return false;
+                    return true;
+                }
+            }).collect(Collectors.toList());
+
+            Integer overallAmount = 0;
+            for(ProductPosition pos: productPositions){
+                overallAmount += pos.getCurrentAmount();
+            }
+            if(overallAmount < requestedAmount){
+                productsAvailableAmounts.put(productId, overallAmount);
+            }
+
+            // product price count
+            Product product = productRepo.findById(productId).get();
```

Comment:
Зачем идти в репо, когда продукт у тебя под рукой. Если нам хватает позиций, то по-любому как минимум одна позиция в листе должна быть, следовательно:
```suggestion
            Product product = productPositions.get(0).getProduct();
```